### PR TITLE
[fix] NetexIDF: only 1 active day not skipped

### PR DESCRIPTION
**Pull request by datanel**
- only one active day not skipped (it was skipped id `properties` node was not present)
- order day type assignments 
    - Operation days
    - Active days
    - Inactive days

### CODE REVIEW

File: src/netex_idf/calendars.rs [hunk 1]
```diff
@@ -41,19 +41,43 @@ struct ValidityPatternIterator<'a> {
...
+        }
+
+        // operating periods are treated first
+        // then active days
+        // then inactive days
+        // to ensure that inactive days are removed if they appears before
+        // operating periods in the XML feed.
+        dtas.extend(ops);
+        dtas.extend(actives);
+        dtas.extend(inactives);
```
Comment (woshilapin) [inline src/netex_idf/calendars.rs]: What is done here is basically ordering objects of type `DayTypeAssignment`. How about implementing `PartialOrd` and `Ord` instead? I gave it a try in [this PR](https://github.com/datanel/transit_model/pull/3) and it works pretty well :smile:

File: src/netex_idf/calendars.rs [hunk 2]
```diff
@@ -507,5 +536,50 @@ mod tests {
...
+
+        #[test]
+        fn inactive_day_before_operating_period() {
+            let validity_period = ValidityPeriod {
+                start_date: Date::from_ymd(2019, 6, 1),
+                end_date: Date::from_ymd(2019, 7, 31),
+            };
+            let weekday_pattern = vec![Weekday::Sat, Weekday::Sun].into_iter().collect();
+            let operating_period = ValidityPeriod {
+                start_date: Date::from_ymd(2019, 7, 1),
```
Comment (woshilapin) [inline src/netex_idf/calendars.rs]: How about moving the start date to the 6th. This test is really about testing the inactive day, not the week pattern.
```suggestion
                start_date: Date::from_ymd(2019, 7, 6),
```

File: tests/fixtures/netexidf2ntfs/output/stops.txt [hunk 3]
```diff
@@ -2,12 +2,12 @@ stop_id,stop_name,visible,fare_zone_id,stop_lon,stop_lat,location_type,parent_st
 prefix:FR::Quay:50144192:FR1,MAISON DU CITOYEN,1,3,2.476489952100393,48.84657571536898,0,prefix:FR::multimodalStopPlace:stop_area_1:FR1,Europe/Paris,,,,
 prefix:FR::Quay:with_zder_not_found:FR1,République,1,,2.3633222823037054,48.86766535188486,0,prefix:Navitia:FR::Quay:with_zder_not_found:FR1,Europe/Paris,,prefix:1,,
 prefix:FR::Quay:50068421:FR1,Avenue de la Gare,1,,2.695967813434811,49.03392183291665,0,prefix:FR::monomodalStopPlace:sp_with_parent_not_found:FR1,Europe/Paris,,prefix:2,,
+prefix:FR::Quay:50114580:FR1,BIBLIOTHEQUE NATIONALE,1,1,2.337949003337833,48.8679593749302,0,prefix:Navitia:FR::Quay:50114580:FR1,Europe/Paris,,prefix:3,,
```
Comment (woshilapin) [inline tests/fixtures/netexidf2ntfs/output/stops.txt]: Nothing changed in this file if I look closely. Why the difference then?

File: tests/fixtures/netexidf2ntfs/output/comment_links.txt [hunk 4]
```diff
@@ -1,3 +1,3 @@
 object_id,object_type,comment_id
-prefix:stif:VehicleJourney:local-26-C00114-1:LOC,trip,prefix:STIF:Notice:C00164_7:LOC
```
Comment (woshilapin) [inline tests/fixtures/netexidf2ntfs/output/comment_links.txt]: Nothing changed in this file, why the difference then?

**Comment by woshilapin** [comment]
Nice new tests to cover the new use cases.

File: src/netex_idf/calendars.rs [hunk 5]
```diff
@@ -65,20 +89,21 @@ impl<'a> Iterator for ValidityPatternIterator<'a> {
         if self.current_date > self.end_date {
             return None;
         }
-        if let Some(dtas) = self.day_type_assignments {
-            let is_included = dtas.iter().fold(false, |is_included, dta| match dta {
```
[the remark this answers is not in the document]
Comment (datanel) [reply src/netex_idf/calendars.rs]: It just doesn't work beacause of borrow checker

File: src/netex_idf/calendars.rs [hunk 6]
```diff
@@ -507,5 +536,50 @@ mod tests {
...
             assert!(validity_pattern_iterator.next().is_none());
         }
+
+        #[test]
+        fn only_one_active_day() {
+            let validity_period = ValidityPeriod {
+                start_date: Date::from_ymd(2019, 6, 1),
+                end_date: Date::from_ymd(2019, 7, 31),
+            };
+            let weekday_pattern = vec![Weekday::Sat, Weekday::Sun].into_iter().collect();
```
Comment (ArnaudOggy) [inline src/netex_idf/calendars.rs]: let weekday_pattern = HashSet::new();

File: tests/fixtures/netexidf2ntfs/output/comment_links.txt (hunk 4 above)
Comment (datanel) [reply tests/fixtures/netexidf2ntfs/output/comment_links.txt]: nothing, I just execute the binary on the fixtures and I know there is a BtreeSet that reorder lines. I'll change it.

File: src/netex_idf/calendars.rs (hunk 1 above)
Comment (datanel) [reply src/netex_idf/calendars.rs]: it seems complicated but more idiomatic

File: src/netex_idf/calendars.rs (hunk 2 above)
Comment (datanel) [reply src/netex_idf/calendars.rs]: I can change it but I "test" implicitly the weekly pattern as well.

Comment (woshilapin) [reply src/netex_idf/calendars.rs]: It's a unit test. Other tests already are taking care of testing the week pattern, that's why I mentioned it.

File: src/netex_idf/calendars.rs [hunk 7]
```diff
@@ -24,30 +24,59 @@ use chrono::{Datelike, NaiveDateTime, Weekday};
...
+// operating periods in the XML feed.
+impl PartialOrd for DayTypeAssignment<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use DayTypeAssignment::*;
+        match (self, other) {
+            (OperatingPeriod(op1), OperatingPeriod(op2)) => {
+                op1.start_date.partial_cmp(&op2.start_date)
+            }
+            (ActiveDay(active1), ActiveDay(active2)) => active1.partial_cmp(&active2),
+            (OperatingPeriod(_), _) | (ActiveDay(_), _) => Some(Ordering::Less),
```
Comment (woshilapin) [inline src/netex_idf/calendars.rs]: This factorization is incorrect. It wouldn't be correct if `self` is of type `ActiveDay` and `other` is of type `OperatingPeriod`: you would expect `Ordering::Greater` but the actual code will return `Ordering::Less` which is a mistake.

```rust
#[test]  
// This test is failing                                                               
fn ord() {                                                              
    let dta1 = DayTypeAssignment::ActiveDay(Date::from_ymd(2020, 1, 1));
    let validity_period = ValidityPeriod {                              
        start_date: Date::from_ymd(2020, 1, 1),                         
        end_date: Date::from_ymd(2020, 1, 1),                           
    };                                                                  
    let dta2 = DayTypeAssignment::OperatingPeriod(&validity_period);    
    assert_eq!(Ordering::Greater, dta1.cmp(&dta2));                     
}                                                                       
```

The condition `(OperatingPeriod(_), _)` must be before `(ActiveDay(a1), ActiveDay(a2))`.

Comment (datanel) [reply src/netex_idf/calendars.rs]: fixed

**Pull request merged**